### ISSUE TITLE
Support for PEX

### ISSUE THREAD

**Issue by espenjutte** [Stefan-Code/peer-harvest#5]
Collection of IPs is currently only done over DHT and tracker scrape. Scrape nets 50-200 IPs due to hard tracker limits and DHT nets only a small portion of the network. 

To get a proper view of the network PEX support is needed. 

PEX support could be added by using either for example bittorrent-swarm module (https://github.com/feross/bittorrent-swarm) with ut_pex (https://github.com/fisch0920/ut_pex).

**Comment by espenjutte** [comment]
I'm working on a prototype with PEX support right now, when i figure out Git i'l try to properly fork it.

**Comment by Stefan-Code** [comment]
I created a [seperate branch](https://github.com/Stefan-Code/peer-harvest/tree/PEX) for working on it.

**Comment by Stefan-Code** [comment]
Now I remember why I didn't implement it while creating this Program: 

```
A peer exchange message should not be sent more frequently than once a minute.
[...]
Some clients may choose to enforce these limits and drop connections which don't obey these limits.
```

[Source](https://wiki.theory.org/BitTorrentPeerExchangeConventions)
That makes discovery via PEX painfully slow for IP gathering.

**Comment by espenjutte** [comment]
Sure, for a single client. But you can always ask more peers for more PEX info (which is what i'm doing). So basically only ask each client once disconnect and move on, which should obey that.

**Comment by Stefan-Code** [comment]
Ah well, yup that seems to work...

**Comment by espenjutte** [comment]
So i'm thinking the following:
1: PEX connect to peers we get from the tracker and from DHT.
2: ask for PEX-info
3: When we receive the PEX-list -> add the new PEX-clients to the peer-connection queue
4: drop connection to the peer
5: repeat 2 for new peers

Problem here is of course: when do we stop? We currently have a timeout running, What i'm thinking might be more effective would be to have a configurable X-second decaying timeout for new peers. In other words: If we don't see any new peers for 10 seconds, end the script and dump the file.

Another option would be to keep polling trackers, DHT and PEX until the script ends. This is a bit less brute-force but we would have to be mindfull of timeouts and kinda manage our connections/who to re-poll etc. This could work since PEX-connection actually send diffs, (added/removed) since last request, so one could in theory keep polling to get a overview of the network.

**Comment by Stefan-Code** [comment]
+1 for a decaying timeout, that's a great idea. You could open a seperate issue for that. Combining it with the existing timeout would be possible, too. I fixed the issue with the duplicate IP adresses btw, it's currently in the PEX branch, although I think I'll merge it with the master before continuing working on PEX.

**Comment by Stefan-Code** [comment]
@espenjutte any progress on the PEX issue? I fixed pretty much all the others now and just wanted to know if you have any new code to share before starting to work on it. Check out the latest release (or just the master branch) for the newly available features (like --timeout-no-peers).

**Comment by espenjutte** [comment]
Nay, dont have time for code until the 13th, so the proof of concept is what I have for now sadly.

**Comment by Stefan-Code** [comment]
Fixed here: https://github.com/Stefan-Code/peer-harvest/commit/39eae05a6a99869ce7fff381364563c51e7425ca in PEX branch. Will merge to master after some testing.

**Comment by Stefan-Code** [comment]
This fix is now in the master branch (version 1.1.0). Still needs some testing though, but it seems to work pretty well.

**Comment by espenjutte** [comment]
That's amazing @Stefan-Code. I'm quickly testing it now and then look at it fully after the 13th.
